File: Data Structures & Algorithms/sliding-window-maximum/submission-1.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> deck = new ArrayDeque<>();
        int[] ret = new int[nums.length - k + 1];
        int idx = 0;

        for (int i = 0; i < nums.length; i++) {
            while (!deck.isEmpty() && deck.peekFirst() <= i - k) {
                deck.removeFirst();
            }
            while (!deck.isEmpty() && nums[deck.peekLast()] <= nums[i]) {
                deck.removeLast();
            }
            deck.addLast(i);
            if (i >= k - 1) {
                ret[idx] = nums[deck.peekFirst()];
                idx++;
            }

        }

        return ret;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pref = [1] * len(nums)
        suff = [1] * len(nums)
        ret = [0] * len(nums)
        for i in range(1, len(nums)):
            pref[i] = nums[i-1] * pref[i-1]
        for i in range(len(nums) - 2, -1, -1):
            suff[i] = nums[i+1] * suff[i+1]
        for i in range(len(pref)):
            ret[i] = pref[i] * suff[i]
        return ret

        